src/Vira/Supervisor.hs

```diff
@@ -103,43 +105,61 @@ startTask' taskId pwd h = runProcs . toList
+          h $ Left e
+          throwIO e
 
     runProc :: CreateProcess -> Eff es (Maybe Pid, ExitCode)
     runProc proc = do
       log Debug $ "Starting task: " <> show (cmdspec proc)
       logToWorkspaceOutput taskId pwd $ "Starting task: " <> show (cmdspec proc)
       outputHandle <- openFile (outputLogFile pwd) AppendMode
-      let processSettings =
-            Process.alwaysUnderPath pwd
-              >>> Process.redirectOutputTo outputHandle
-      (_, _, _, ph) <- createProcess $ proc & processSettings
-      pid <- getPid ph
-      log Debug $ "Task spawned (pid=" <> show pid <> "): " <> show (cmdspec proc)
-      exitCode <- waitForProcess ph
-      log Debug $ "Task finished (pid=" <> show pid <> "): " <> show (cmdspec proc)
-      hClose outputHandle
-      logToWorkspaceOutput taskId pwd $ "A task (pid=" <> show pid <> ") finished with exit code " <> show exitCode
-      log Debug "Workspace log done"
-      pure (pid, exitCode)
+      flip finally (hClose outputHandle) $ do
+        let processSettings =
+              Process.alwaysUnderPath pwd
+                >>> Process.redirectOutputTo outputHandle
+        (_, _, _, ph) <- createProcess $ proc & processSettings
+        pid <- getPid ph
+        log Debug $ "Task spawned (pid=" <> show pid <> "): " <> show (cmdspec proc)
+
+        exitCode <-
+          -- `mask` ensures proper cleanup after `UserKilled`, by protecting from further asynchronous interruptions
+          mask $ \restore ->
+            restore (waitForProcess ph)
+              `catch` \(exc :: TaskException) -> do
+                -- This handler runs with async exceptions masked.
+                case exc of
+                  UserKilled -> do
+                    log Info $ "Terminating process (pid=" <> show pid <> ") for task " <> show taskId
+                    terminateProcess ph `catch` \(e :: SomeException) ->
+                      log Error $ "Failed to terminate process " <> show pid <> " for task " <> show taskId <> ": " <> show e
+                    -- Reap the process to prevent from becoming zombie
+                    _ <- waitForProcess ph
+                    throwIO exc -- re-trhwo `UserKilled` to ensure the task status is updated to `Killed`
+        log Debug $ "Task finished (pid=" <> show pid <> "): " <> show (cmdspec proc)
+        logToWorkspaceOutput taskId pwd $ "A task (pid=" <> show pid <> ") finished with exit code " <> show exitCode
+        log Debug "Workspace log done"
+        pure (pid, exitCode)
 
--- | Kill a task
-killTask :: TaskSupervisor -> TaskId -> Eff App.AppStack ()
+-- | Kill an active task
+killTask :: (Concurrent :> es, Log Message :> es, IOE :> es) => TaskSupervisor -> TaskId -> Eff es ()
 killTask supervisor taskId = do
-  log Info $ "Killing task " <> show taskId
   modifyMVar_ (tasks supervisor) $ \tasks -> do
-    for_ (Map.lookup taskId tasks) $ \Task {..} ->
-      cancel asyncHandle
+    for_ (Map.lookup taskId tasks) $ \Task {..} -> do
```

Comment:
`for_` is not idiomatic for iterating something that's not a collection (`Maybe` here). You want to use [`whenJust`](https://hackage.haskell.org/package/relude-1.2.2.0/docs/Relude-Monad-Maybe.html#v:whenJust). More generally, I recommend familiarizing yourself with the `relude` library.